tsdb/head_append.go

```diff
@@ -978,23 +978,39 @@ func exemplarsForEncoding(es []exemplarWithSeriesRef) []record.RefExemplar {
 	return ret
 }
 
-// Commit writes to the WAL and adds the data to the Head.
-// TODO(codesome): Refactor this method to reduce indentation and make it more readable.
-func (a *headAppender) Commit() (err error) {
-	if a.closed {
-		return ErrAppenderClosed
-	}
-	defer func() { a.closed = true }()
-
-	if err := a.log(); err != nil {
-		_ = a.Rollback() // Most likely the same error will happen again.
-		return fmt.Errorf("write to WAL: %w", err)
-	}
-
-	if a.head.writeNotified != nil {
-		a.head.writeNotified.Notify()
-	}
+type appenderCommitContext struct {
```

Comment:
I don't know enough about the go compiler to know if it can use registers to hold these values now. With local variables I assume the compiler can optimize usage, but these are now written to memory in the struct, so that might introduce considerable overhead. 

I think we definitely need a benchmark compare to original.